tools/runtime-codegen/src/main.rs

```diff
@@ -0,0 +1,173 @@
+struct Command {
+	#[clap(name = "from-node-url", long, value_parser)]
+	node_url: Option<Uri>,
+	#[clap(name = "from-wasm-file", long, value_parser)]
+	wasm_file: Option<String>,
+}
+
+enum RuntimeMetadataSource {
+	NodeUrl(Uri),
+	WasmFile(wasm_loader::Source),
+}
+
+impl RuntimeMetadataSource {
+	fn from_command(cmd: Command) -> color_eyre::Result<Self> {
+		match (cmd.node_url, cmd.wasm_file) {
+			(Some(_), Some(_)) => Err(eyre::eyre!(
+				"Please specify one of `--from-node-url` or `--from-wasm-file` but not both"
+			)),
+			(None, None) =>
+				Err(eyre::eyre!("Please specify one of `--from-node-url` or `--from-wasm-file`")),
+			(Some(node_url), None) => Ok(Self::NodeUrl(node_url)),
+			(None, Some(source)) =>
+				Ok(Self::WasmFile(wasm_loader::Source::File(PathBuf::from(source)))),
+		}
+	}
+}
+
+struct TypeSubstitute {
+	subxt_type: syn::Path,
+	substitute: syn::Path,
+}
+
+impl TypeSubstitute {
+	fn simple(subxt_type: &str) -> Self {
+		Self {
+			subxt_type: syn::parse_str::<syn::Path>(subxt_type).unwrap(),
+			substitute: syn::parse_str::<syn::Path>(&format!("::{}", subxt_type)).unwrap(),
+		}
+	}
+
+	fn custom(subxt_type: &str, substitute: &str) -> Self {
+		Self {
+			subxt_type: syn::parse_str::<syn::Path>(subxt_type).unwrap(),
+			substitute: syn::parse_str::<syn::Path>(substitute).unwrap(),
+		}
+	}
+}
+
+fn print_runtime(runtime_api: proc_macro2::TokenStream) {
+	println!(
+		"// Copyright 2019-2023 Parity Technologies (UK) Ltd.
+		// This file is part of Parity Bridges Common.
+
+		// Parity Bridges Common is free software: you can redistribute it and/or modify
+		// it under the terms of the GNU General Public License as published by
+		// the Free Software Foundation, either version 3 of the License, or
+		// (at your option) any later version.
+
+		// Parity Bridges Common is distributed in the hope that it will be useful,
+		// but WITHOUT ANY WARRANTY; without even the implied warranty of
+		// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+		// GNU General Public License for more details.
+
+		// You should have received a copy of the GNU General Public License
+		// along with Parity Bridges Common.  If not, see <http://www.gnu.org/licenses/>.
+
+		//! Autogenerated runtime API
+		//! THIS FILE WAS AUTOGENERATED USING parity-bridges-common::runtime-codegen
+		//! EXECUTED COMMAND: {}
+
+		{}
+		",
+		env::args().collect::<Vec<String>>().join(" "),
+		runtime_api
+	);
+}
+
+fn main() -> color_eyre::Result<()> {
+	let args: Command = Command::parse();
+	let metadata_source = RuntimeMetadataSource::from_command(args)?;
+
+	// Module under which the API is generated.
+	let item_mod = syn::parse_quote!(
+		pub mod api {}
+	);
+	// Default module derivatives.
+	let mut derives = DerivesRegistry::new(&CratePath::default());
+	derives.extend_for_all(vec![syn::parse_quote!(Clone)]);
+	// Type substitutes
+	let mut type_substitutes = TypeSubstitutes::new(&CratePath::default());
+	type_substitutes.extend(
```

Comment:
Not an issue - just a question. I can see that the generated code has something like `mod sp_runtime`. That's why we need those substitutions, right? But why only for several types? Are they types that we use in relay?